pkg/controller/plugins.go

```diff
@@ -0,0 +1,228 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/util"
+	"k8s.io/kubernetes/pkg/util/errors"
+
+	"github.com/golang/glog"
+)
+
+// Plugin is an interface to controller plugins that can be run
+// in kube-controller-manager.  A PluginMgr handles instantiating
+// and managing the lifecycle of controllers.
+type Plugin interface {
+	// Init initializes the plugin.  This will be called once before Run.
+	Init(host Host)
+
+	// Name returns the plugin's name.  Plugins should use namespaced names
+	// such as "example.com/replication-controller".  The "kubernetes.io" namespace is
+	// reserved for plugins which are bundled with kubernetes.
+	Name() string
+
+	// NewController returns an instance of this plugin's controller.
+	NewController() (Controller, error)
```

Comment:
ok, so keep the Status func on a Controller.

As for init/start, I can see why Volumes would need init (initialized all plugins once, we'll get many different instances of them later).

Controllers, though, are just instantiated once and run.  

Do we need separate Plugin and Controller interfaces, then?  

instead of plugin.NewController().Start() is it just plugin.Start()?